required/latex-lab/latex-lab-everymath.dtx

```diff
@@ -0,0 +1,347 @@
+%
+% This file implements capture of all math mode material at the top level.
+% It provides one code-level interface to use this captured input.
+%
+% \begin{function}{\math_processor:n}
+%   \begin{syntax}
+%     \cs{math_processor:n} \Arg{tokens}
+%   \end{syntax}
+%   Declares that the captured math content should be passed to the
+%   \meta{tokens}, which will receive the content as |#1|.
+% \end{function}
+%
+% \MaybeStop{\setlength\IndexMin{200pt}  \PrintIndex  }
+%
+% \section{The Implementation}
+%
+%    \begin{macrocode}
+%<@@=math>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<*kernel>
+%    \end{macrocode}
+%
+% \subsection{File declaration}
+%    \begin{macrocode}
+\ProvidesFile{latex-lab-everymath.ltx}
+        [2022-12-09 v0.1a Grab all the math(s)]
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\ExplSyntaxOn
+%    \end{macrocode}
+%
+% \subsection{Setup}
+%
+% Loading \pkg{amsmath} is an absolute requirement: this avoids needing to
+% have conditional definitions and deals with how to define \cs{[}/\cs{]}
+% neatly.
+%    \begin{macrocode}
+\AddToHook { begindocument / before }
+  { \RequirePackage { amsmath } }
+%    \end{macrocode} 
+%
+% \subsection{Data structures}
+%
+% \begin{variable}{\l_@@_cmd_bool}
+%   Tracks whether we are dealing with \LaTeX{}'s math mode (which is
+%   referred to here as \texttt{cmd}).
+%    \begin{macrocode}
+\bool_new:N \l_@@_cmd_bool
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{variable}{\l_@@_inner_bool}
+%   Tracks if we are inside a top-level inline math context.
+%    \begin{macrocode}
+\bool_new:N \l_@@_inner_bool
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{variable}{\l_@@_display_bool}
+%   Tracks display style (cases where \cs{everydisplay} is triggered).
+%    \begin{macrocode}
+\bool_new:N \l_@@_display_bool
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{variable}{\l_@@_env_bool}
+%   Tracks math environments which deal with grabbing themselves.
+%    \begin{macrocode}
+\bool_new:N \l_@@_env_bool
+%    \end{macrocode}
+% \end{variable}
+%
+% \subsection{Interface commands}
+%
+% \begin{macro}{\@@_process:n, \@@_process_aux:n}
+%   A no-op place-holder; the internal wrapper means that it does not need to
+%   be concerned with internals.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_process:n #1
+  {
+    \legacy_if:nF { measuring@ }
+      { \@@_process_aux:n {#1} }
+  }
+\cs_new_protected:Npn \@@_process_aux:n #1 { }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\math_processor:n}
+%   A simple installer
+%    \begin{macrocode}
+\cs_new_protected:Npn \math_processor:n #1
+  { \cs_set_protected:Npn \@@_process_aux:n ##1 {#1} }
+%    \end{macrocode}
+% \end{macro}
+%
+% \subsection{Content grabbing}
+%
+% \begin{macro}{\@@_grab_inline_dollar:w}
+%   Grab up to a single |$|, for inline math mode.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_inline_dollar:w % $
+  #1 $
+  {
+    \@@_process:n {#1} % $
+    #1 $
+  }
```

Comment:
what happens if the  closing dollar is hidden, e.g., an environment that has the dollars in begin+end?
We can't really avoid that, but perhaps we need to offer an interface (maybe a flag to set?) so that such cases can be handled if the code is updated. --- I realize I write comments while reading along, so maybe that is offered later ...

Reply:
Well really what we need is all math mode to be verb-like, as is already done by `amsmath`. I've not exposed the booleans at the moment, but we could. I think it's pretty unlikely to be required for inline maths.